trio/_core/_tests/test_instrumentation.py

```diff
@@ -1,40 +1,45 @@
+from __future__ import annotations
+
+from typing import Container, Iterable, NoReturn
+
 import attr
 import pytest
 
 from ... import _abc, _core
+from ...lowlevel import Task
 from .tutil import check_sequence_matches
 
 
 @attr.s(eq=False, hash=False)
-class TaskRecorder:
-    record = attr.ib(factory=list)
+class TaskRecorder(_abc.Instrument):
+    record: list[tuple[str, Task | None]] = attr.ib(factory=list)
```

Comment:
I think you can keep the variable length tuples and just use tuple length to differentiate (I believe that narrows in mypy? If not, ignore this comment!)

Specifically `list[tuple[str] | tuple[str, Task]]`

Reply:
It does as of last week - https://github.com/python/mypy/pull/16237. So not quite yet.

Reply:
I'll leave it as is then.